### PR TITLE
Declare post collection params so they can be exposed

### PR DESCRIPTION
**Pull request by danielbachhuber**
See #924, #555

### CODE REVIEW

File: lib/endpoints/class-wp-rest-posts-controller.php [hunk 1]
```diff
@@ -35,11 +35,12 @@ public function register_routes() {
 			}
```
Comment (danielbachhuber) [inline lib/endpoints/class-wp-rest-posts-controller.php]: @WP-API/amigos What should I do with legacy `get_allowed_query_vars()` ?

**Comment by danielbachhuber** [comment]
@WP-API/amigos I'm not 100% sure what we decided to do with `WP_Query`-style arguments (e.g. also supporting `posts_per_page` in addition to `page`).

Given we don't support `WP_User_Query`-style arguments et al, and trying to magically support them would be a pretty massive hack, I think we should drop them for now.

**Comment by rachelbaker** [comment]
> I'm not 100% sure what we decided to do with WP_Query-style arguments 

I don't know if we discussed this, but it sounds like we should.

File: lib/endpoints/class-wp-rest-posts-controller.php (hunk 1 above)
Comment (rachelbaker) [inline lib/endpoints/class-wp-rest-posts-controller.php]: The API has always supported the public query_vars of WP_Query, included a few others, and provided a filter to add support for additional vars.  There are some public query vars that are pretty pointless (example `withcomments`) but trying to normalize all these params sounds like a headache to me.

**Comment by rachelbaker** [comment]
Looks like this needs some more discussion.  @rmccue @joehoyle thoughts?

**Comment by joehoyle** [comment]
When we discussed this at LoopConf, I was under the impression we settled on the following:

1. Create new query params, for example `page`, `per_page`, `slug` etc.
2. Put all current WP_Query style WordPress query names under a `filter` or `wp_query` argument. So these would be equivalent: 

`GET /wp/v2/posts?per_page=10`

`GET /wp/v2/posts?wp_query[posts_per_page]=10`

That way, we can support WordPress devs thinking the a `WP_Query` mindset, and give them the ability to construct more complex queries (though ofcourse of the public query vars that we currently use). Then have a clean top level args that will be consistent across all collections.

**Comment by danielbachhuber** [comment]
> `GET /wp/v2/posts?wp_query[posts_per_page]=10`

I dig this — seems like a reasonable compromise.

**Comment by rmccue** [comment]
If we're going to do it, keep the `filter` name so it's consistent with v1. That way, v2 = v1 + additional better named querying, rather than completely different.

**Comment by danielbachhuber** [comment]
I dig `filter`

**Comment by rmccue** [comment]
Marking as Needs Refresh, since `filter` needs to be added in.

**Comment by danielbachhuber** [comment]
Closing in favor of #1793

**Pull request closed without merging**